cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java

```diff
@@ -70,15 +74,30 @@ public void updateLaunchArguments(ProgramSchedule schedule, List<Notification> n
     return unitTriggers;
   }
 
+  /**
+   * Get all inner triggers as SatisfiableTrigger's in this trigger. Cannot initialize satisfiableTriggers in
+   * the constructor because the constructor will not be called if this is instantiated by {@link TriggerCodec}.
+   */
+  public List<SatisfiableTrigger> getSatisfiableTriggers() {
+    if (satisfiableTriggers == null) {
+      satisfiableTriggers = new ArrayList<>();
+      for (Trigger trigger : getTriggers()) {
+        // the trigger is guaranteed to be an instance of SatisfiableTrigger because of the parameter type of
```

Comment:
when you do stuff like this, it tightly couples multiple classes, making things really easy to break. A change somewhere else will result in a really confusing ClassCastException here.

Reply:
could you make the superclass generic (`T extends Trigger`) so that getTriggers() returns a `List<T>`?